File: src/LongleyRice.cpp
#include "..\include\itm.h"
#include "..\include\Enums.h"
#include "..\include\Errors.h"
#include "..\include\Warnings.h"

/*=============================================================================
 |
 |  Description:  Compute the reference attenuation, using the 
 |                Longley-Rice method
 |
 |        Input:  theta_hzn[2]      - Terminal horizon angles
 |                f__mhz            - Frequency, in MHz
 |                Z_g               - Complex surface transfer impedance
 |                d_hzn__meter[2]   - Terminal horizon distances, in meters
 |                h_e__meter[2]     - Effective terminal heights, in meters
 |                gamma_e           - Curvature of the effective earth
 |                N_s               - Surface refractivity, in N-Units
 |                delta_h__meter    - Terrain irregularity parameter
 |                h__meter[2]       - Terminal structural heights, in meters
 |                d__meter          - Path distance, in meters
 |                mode              - Mode of operation (P2P or Area)
 |
 |      Outputs:  A_ref__db         - Reference attenuation, in dB
 |                warnings          - Warning flags
 |                propmode          - Mode of propagation value
 |
 |      Returns:  error             - Error code
 |
 *===========================================================================*/
int LongleyRice(double theta_hzn[2], double f__mhz, complex<double> Z_g, double d_hzn__meter[2], 
    double h_e__meter[2], double gamma_e, double N_s, double delta_h__meter, double h__meter[2], 
    double d__meter, int mode, double *A_ref__db, long *warnings, int *propmode)
{
    double d_x__meter = 0;

    // effective earth radius
    double a_e__meter = 1 / gamma_e;

    double d_hzn_s__meter[2];
    // Terrestrial smooth earth horizon distance approximation
    for (int i = 0; i < 2; i++)
        d_hzn_s__meter[i] = sqrt(2.0 * h_e__meter[i] * a_e__meter);

    // Maximum line-of-sight distance for smooth earth
    double d_sML__meter = d_hzn_s__meter[0] + d_hzn_s__meter[1];

    // Maximum line-of-sight distance for actual path
    double d_ML__meter = d_hzn__meter[0] + d_hzn__meter[1];

    // Angular distance of line-of-sight region
    double theta_los = -MAX(theta_hzn[0] + theta_hzn[1], -d_ML__meter / a_e__meter);

    // Check validity of small angle approximation
    if (abs(theta_hzn[0]) > 200e-3)
        *warnings |= WARN__TX_HORIZON_ANGLE;
    if (abs(theta_hzn[1]) > 200e-3)
        *warnings |= WARN__RX_HORIZON_ANGLE;

    // Checks that the actual horizon distance can't be less than 1/10 of the smooth earth horizon distance
    if (d_hzn__meter[0] < 0.1 * d_hzn_s__meter[0])
        *warnings |= WARN__TX_HORIZON_DISTANCE_1;
    if (d_hzn__meter[1] < 0.1 * d_hzn_s__meter[1])
        *warnings |= WARN__RX_HORIZON_DISTANCE_1;

    // Checks that the actual horizon distance can't be greater than 3 times the smooth earth horizon distance
    if (d_hzn__meter[0] > 3.0 * d_hzn_s__meter[0])
        *warnings |= WARN__TX_HORIZON_DISTANCE_2;
    if (d_hzn__meter[1] > 3.0 * d_hzn_s__meter[1])
        *warnings |= WARN__RX_HORIZON_DISTANCE_2;

    // Check the surface refractivity
    if (N_s < 150)
        return ERROR__SURFACE_REFRACTIVITY_SMALL;
    if (N_s > 400)
        return ERROR__SURFACE_REFRACTIVITY_LARGE;
    if (N_s < 250) // 150 <= N_s < 250
        *warnings |= WARN__SURFACE_REFRACTIVITY;

    // Check effective earth size
    if (a_e__meter < 4000000 || a_e__meter > 13333333)
        return ERROR__EFFECTIVE_EARTH;

    // Check ground impedance
    if (Z_g.real() <= abs(Z_g.imag()))
        return ERROR__GROUND_IMPEDANCE;

    // Select two distances far in the diffraction region
    double d_3__meter = MAX(d_sML__meter, d_ML__meter + 5.0 * pow(pow(a_e__meter, 2) / f__mhz, 1.0 / 3.0));
    double d_4__meter = d_3__meter + 10.0 * pow(pow(a_e__meter, 2) / f__mhz, 1.0 / 3.0);

    // Compute the diffraction loss at the two distances
    double A_3__db = DiffractionLoss(d_3__meter, d_hzn__meter, h_e__meter, Z_g, a_e__meter, delta_h__meter, h__meter, mode, theta_los, d_sML__meter, f__mhz);
    double A_4__db = DiffractionLoss(d_4__meter, d_hzn__meter, h_e__meter, Z_g, a_e__meter, delta_h__meter, h__meter, mode, theta_los, d_sML__meter, f__mhz);

    // Compute the slope and intercept of the diffraction line
    double M_d = (A_4__db - A_3__db) / (d_4__meter - d_3__meter);
    double A_d0__db = A_3__db - M_d * d_3__meter;

    double d_min__meter = abs(h_e__meter[0] - h_e__meter[1]) / 200e-3;

    if (d__meter < d_min__meter)
        *warnings |= WARN__PATH_DISTANCE_TOO_SMALL_1;
    if (d__meter < 1e3)
        *warnings |= WARN__PATH_DISTANCE_TOO_SMALL_2;
    if (d__meter > 1000e3)
        *warnings |= WARN__PATH_DISTANCE_TOO_BIG_1;
    if (d__meter > 2000e3)
        *warnings |= WARN__PATH_DISTANCE_TOO_BIG_2;

    // if the path distance is less than the maximum smooth earth line of sight distance...
    if (d__meter < d_sML__meter)
    {
        // Compute the diffraction loss at the maximum smooth earth line of sight distance
        double A_sML__db = d_sML__meter * M_d + A_d0__db;

        // [ERL 79-ITS 67, Eqn 3.16a], in meters instead of km and with MIN() part below
        double d_0__meter = 0.04 * f__mhz * h_e__meter[0] * h_e__meter[1];

        double d_1__meter;
        if (A_d0__db >= 0.0)
        {
            d_0__meter = MIN(d_0__meter, 0.5 * d_ML__meter);                // other part of [ERL 79-ITS 67, Eqn 3.16a]
            d_1__meter = d_0__meter + 0.25 * (d_ML__meter - d_0__meter);    // [ERL 79-ITS 67, Eqn 3.16d]
        }
        else
            d_1__meter = MAX(-A_d0__db / M_d, 0.25 * d_ML__meter);

        double A_1__db = LineOfSightLoss(d_1__meter, h_e__meter, Z_g, delta_h__meter, M_d, A_d0__db, d_sML__meter, f__mhz);

        bool flag = false;

        double kHat_1__db_per_meter = 0;
        double kHat_2__db_per_meter = 0;

        if (d_0__meter < d_1__meter)
        {
            double A_0__db = LineOfSightLoss(d_0__meter, h_e__meter, Z_g, delta_h__meter, M_d, A_d0__db, d_sML__meter, f__mhz);

            double q = log(d_sML__meter / d_0__meter);

            // [ERL 79-ITS 67, Eqn 3.20]
            kHat_2__db_per_meter = MAX(0.0, ((d_sML__meter - d_0__meter) * (A_1__db - A_0__db) - (d_1__meter - d_0__meter) * (A_sML__db - A_0__db)) / ((d_sML__meter - d_0__meter) * log(d_1__meter / d_0__meter) - (d_1__meter - d_0__meter) * q));

            flag = A_d0__db > 0.0 || kHat_2__db_per_meter > 0.0;

            if (flag)
            {
                // [ERL 79-ITS 67, Eqn 3.21]
                kHat_1__db_per_meter = (A_sML__db - A_0__db - kHat_2__db_per_meter * q) / (d_sML__meter - d_0__meter);

                if (kHat_1__db_per_meter < 0.0)
                {
                    kHat_1__db_per_meter = 0.0;
                    kHat_2__db_per_meter = DIM(A_sML__db, A_0__db) / q;

                    if (kHat_2__db_per_meter == 0.0)
                        kHat_1__db_per_meter = M_d;
                }
            }
        }

        if (!flag)
        {
            kHat_1__db_per_meter = DIM(A_sML__db, A_1__db) / (d_sML__meter - d_1__meter);
            kHat_2__db_per_meter = 0.0;

            if (kHat_1__db_per_meter == 0.0)
                kHat_1__db_per_meter = M_d;
        }

        double A_o__db = A_sML__db - kHat_1__db_per_meter * d_sML__meter - kHat_2__db_per_meter * log(d_sML__meter);

        // [ERL 79-ITS 67, Eqn 3.19]
        *A_ref__db = A_o__db + kHat_1__db_per_meter * d__meter + kHat_2__db_per_meter * log(d__meter);
    }
    else // this is a trans-horizon path
    {
        // select to points far into the troposcatter region
        double d_5__meter = d_ML__meter + 200e3;
        double d_6__meter = d_ML__meter + 400e3;

        // Compute the troposcatter loss at the two distances
        double h0 = -1;
        double A_6__db = TroposcatterLoss(d_6__meter, theta_hzn, d_hzn__meter, h_e__meter, a_e__meter, N_s, f__mhz, theta_los, &h0);
        double A_5__db = TroposcatterLoss(d_5__meter, theta_hzn, d_hzn__meter, h_e__meter, a_e__meter, N_s, f__mhz, theta_los, &h0);

        double M_s, A_s0__db;

        // if we got a reasonable prediction value back...
        if (A_5__db < 1000.0)
        {
            // Compute the slope of the troposcatter line
            M_s = (A_6__db - A_5__db) / 200e3;

            // Find the diffraction-troposcatter transition distance
            d_x__meter = MAX(MAX(d_sML__meter, d_ML__meter + 1.088 * pow(pow(a_e__meter, 2) / f__mhz, 1.0 / 3.0) * log(f__mhz)), (A_5__db - A_d0__db - M_s * d_5__meter) / (M_d - M_s));

            // Compute the intercept of the troposcatter line
            A_s0__db = (M_d - M_s) * d_x__meter + A_d0__db;
        }
        else
        {
            // troposcatter gives no real results - so use diffraction line parameters for tropo line
            M_s = M_d;
            A_s0__db = A_d0__db;
            d_x__meter = 10e6;
        }

        // Determine if its diffraction or troposcatter and compute the loss
        if (d__meter > d_x__meter)
            *A_ref__db = M_s * d__meter + A_s0__db;
        else
            *A_ref__db = M_d * d__meter + A_d0__db;
    }

    // set mode of propagation
    double delta__meter = d__meter - d_ML__meter;
    if (int(delta__meter) < 0)
        *propmode = MODE__LINE_OF_SIGHT;
    else
    {
        if (d__meter <= d_sML__meter || d__meter <= d_x__meter)
            *propmode = (int(delta__meter) == 0) ? MODE__DIFFRACTION_SINGLE_HORIZON : MODE__DIFFRACTION_DOUBLE_HORIZON;
        else
            *propmode = (int(delta__meter) == 0) ? MODE__TROPOSCATTER_SINGLE_HORIZON : MODE__TROPOSCATTER_DOUBLE_HORIZON;
    }

    // Don't allow a negative loss
    *A_ref__db = MAX(*A_ref__db, 0.0);

    return SUCCESS;
}

File: itm-rs/itm/src/longley_rice.rs
use crate::complex::Complex;
use crate::constants::*;
use crate::diffraction_loss::diffraction_loss;
use crate::errors::*;
use crate::line_of_sight_loss::line_of_sight_loss;
use crate::troposcatter_loss::troposcatter_loss;
use crate::warnings::*;

/// Compute the reference attenuation using the Longley-Rice method.
/// Returns `(error_code, a_ref__db, propmode)`.
pub fn longley_rice(
    theta_hzn: [f64; 2],
    f__mhz: f64,
    z_g: Complex,
    d_hzn__meter: [f64; 2],
    h_e__meter: [f64; 2],
    gamma_e: f64,
    n_s: f64,
    delta_h__meter: f64,
    h__meter: [f64; 2],
    d__meter: f64,
    mode: i32,
    warnings: &mut i32,
) -> (i32, f64, i32) {
    let a_e__meter = 1.0 / gamma_e;

    let d_hzn_s = [
        (2.0 * h_e__meter[0] * a_e__meter).sqrt(),
        (2.0 * h_e__meter[1] * a_e__meter).sqrt(),
    ];
    let d_sml = d_hzn_s[0] + d_hzn_s[1];
    let d_ml = d_hzn__meter[0] + d_hzn__meter[1];
    let theta_los = -(theta_hzn[0] + theta_hzn[1]).max(-d_ml / a_e__meter);

    if theta_hzn[0].abs() > 200e-3 {
        *warnings |= WARN__TX_HORIZON_ANGLE;
    }
    if theta_hzn[1].abs() > 200e-3 {
        *warnings |= WARN__RX_HORIZON_ANGLE;
    }
    if d_hzn__meter[0] < 0.1 * d_hzn_s[0] {
        *warnings |= WARN__TX_HORIZON_DISTANCE_1;
    }
    if d_hzn__meter[1] < 0.1 * d_hzn_s[1] {
        *warnings |= WARN__RX_HORIZON_DISTANCE_1;
    }
    if d_hzn__meter[0] > 3.0 * d_hzn_s[0] {
        *warnings |= WARN__TX_HORIZON_DISTANCE_2;
    }
    if d_hzn__meter[1] > 3.0 * d_hzn_s[1] {
        *warnings |= WARN__RX_HORIZON_DISTANCE_2;
    }

    if n_s < 150.0 {
        return (ERROR__SURFACE_REFRACTIVITY_SMALL, 0.0, MODE__NOT_SET);
    }
    if n_s > 400.0 {
        return (ERROR__SURFACE_REFRACTIVITY_LARGE, 0.0, MODE__NOT_SET);
    }
    if n_s < 250.0 {
        *warnings |= WARN__SURFACE_REFRACTIVITY;
    }

    if a_e__meter < 4_000_000.0 || a_e__meter > 13_333_333.0 {
        return (ERROR__EFFECTIVE_EARTH, 0.0, MODE__NOT_SET);
    }

    if z_g.re <= z_g.im.abs() {
        return (ERROR__GROUND_IMPEDANCE, 0.0, MODE__NOT_SET);
    }

    let cbrt_term = (a_e__meter.powi(2) / f__mhz).powf(1.0 / 3.0);
    let d_3 = f64::max(d_sml, d_ml + 5.0 * cbrt_term);
    let d_4 = d_3 + 10.0 * cbrt_term;

    let a_3 = diffraction_loss(
        d_3,
        d_hzn__meter,
        h_e__meter,
        z_g,
        a_e__meter,
        delta_h__meter,
        h__meter,
        mode,
        theta_los,
        d_sml,
        f__mhz,
    );
    let a_4 = diffraction_loss(
        d_4,
        d_hzn__meter,
        h_e__meter,
        z_g,
        a_e__meter,
        delta_h__meter,
        h__meter,
        mode,
        theta_los,
        d_sml,
        f__mhz,
    );

    let m_d = (a_4 - a_3) / (d_4 - d_3);
    let a_d0 = a_3 - m_d * d_3;

    let d_min = (h_e__meter[0] - h_e__meter[1]).abs() / 200e-3;
    if d__meter < d_min {
        *warnings |= WARN__PATH_DISTANCE_TOO_SMALL_1;
    }
    if d__meter < 1_000.0 {
        *warnings |= WARN__PATH_DISTANCE_TOO_SMALL_2;
    }
    if d__meter > 1_000_000.0 {
        *warnings |= WARN__PATH_DISTANCE_TOO_BIG_1;
    }
    if d__meter > 2_000_000.0 {
        *warnings |= WARN__PATH_DISTANCE_TOO_BIG_2;
    }

    let (a_ref, propmode) = if d__meter < d_sml {
        let a_sml = d_sml * m_d + a_d0;
        let mut d_0 = 0.04 * f__mhz * h_e__meter[0] * h_e__meter[1];
        let d_1;

        if a_d0 >= 0.0 {
            d_0 = f64::min(d_0, 0.5 * d_ml);
            d_1 = d_0 + 0.25 * (d_ml - d_0);
        } else {
            d_1 = f64::max(-a_d0 / m_d, 0.25 * d_ml);
        }

        let a_1 = line_of_sight_loss(d_1, h_e__meter, z_g, delta_h__meter, m_d, a_d0, d_sml, f__mhz);

        let mut flag = false;
        let mut k1 = 0.0_f64;
        let mut k2 = 0.0_f64;

        if d_0 < d_1 {
            let a_0 = line_of_sight_loss(d_0, h_e__meter, z_g, delta_h__meter, m_d, a_d0, d_sml, f__mhz);
            let q = (d_sml / d_0).ln();

            k2 = f64::max(
                0.0,
                ((d_sml - d_0) * (a_1 - a_0) - (d_1 - d_0) * (a_sml - a_0))
                    / ((d_sml - d_0) * (d_1 / d_0).ln() - (d_1 - d_0) * q),
            );

            flag = a_d0 > 0.0 || k2 > 0.0;

            if flag {
                k1 = (a_sml - a_0 - k2 * q) / (d_sml - d_0);
                if k1 < 0.0 {
                    k1 = 0.0;
                    k2 = ((a_sml - a_0).max(0.0)) / q;
                    if k2 == 0.0 {
                        k1 = m_d;
                    }
                }
            }
        }

        if !flag {
            k1 = (a_sml - a_1).max(0.0) / (d_sml - d_1);
            k2 = 0.0;
            if k1 == 0.0 {
                k1 = m_d;
            }
        }

        let a_o = a_sml - k1 * d_sml - k2 * d_sml.ln();
        let a_ref = a_o + k1 * d__meter + k2 * d__meter.ln();
        (a_ref, MODE__LINE_OF_SIGHT)
    } else {
        let d_5 = d_ml + 200_000.0;
        let d_6 = d_ml + 400_000.0;

        let mut h0 = -1.0;
        let a_6 = troposcatter_loss(
            d_6, theta_hzn, d_hzn__meter, h_e__meter, a_e__meter, n_s, f__mhz, theta_los, &mut h0,
        );
        let a_5 = troposcatter_loss(
            d_5, theta_hzn, d_hzn__meter, h_e__meter, a_e__meter, n_s, f__mhz, theta_los, &mut h0,
        );

        let (m_s, a_s0, d_x) = if a_5 < 1000.0 {
            let m_s = (a_6 - a_5) / 200_000.0;
            let d_x = f64::max(
                f64::max(
                    d_sml,
                    d_ml + 1.088 * (a_e__meter.powi(2) / f__mhz).powf(1.0 / 3.0) * f__mhz.ln(),
                ),
                (a_5 - a_d0 - m_s * d_5) / (m_d - m_s),
            );
            let a_s0 = (m_d - m_s) * d_x + a_d0;
            (m_s, a_s0, d_x)
        } else {
            (m_d, a_d0, 10_000_000.0)
        };

        if d__meter > d_x {
            (m_s * d__meter + a_s0, MODE__TROPOSCATTER)
        } else {
            (m_d * d__meter + a_d0, MODE__DIFFRACTION)
        }
    };

    (SUCCESS, a_ref.max(0.0), propmode)
}
